### PR TITLE
Extract region

### PR DESCRIPTION
**Pull request by awalter-bnl**
This adds a function to extract out a region form the raw data.

### CODE REVIEW

**Comment by mpmdean** [comment]
Thanks. The code looks completely pristine to my (limited) eyes. Can I come over to discuss face-to-face a bit? 

I would suggest that we eventually make separate objects describing the processing of the different ROIs of the camera. (I certainly agree we should populate these out of the databroker.) Something very roughly along these lines:

process_dict_low_2theta = {'tag': 'low_2theta',
                           'ROI': [slice(175, 1609), slice(1:1751)],
                           'curvature': np.array([0., 0., 0.]),
                           'bins': None}

process_dict_high_2theta = {'tag': 'high_2theta',
                            'ROI': [slice(175:1609), slice(1753:3503)],
                           'curvature': np.array([0., 0., 0.]),
                           'bins': None}

**Comment by awalter-bnl** [comment]
sure I am happy to discuss face to face, I am still working through the Travis error on this though !-).

File: sixtools/preliminary_processing.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        containing the data from 'region'.
+
+    Examples
+    --------
+    This is intended to be used in the following, one line, call with XXXX
+    being a scan_id or equivalent (start.rixscam.regions is suppose to point
+    to the metadata containing the regions for this scan):
+    >>> extract_region(db[XXXX].data, region, db[XXXX].start.rixscam.regions)
+    '''
+    # step through each of the 'events' ('yields') in raw_data
```
Comment (danielballan) [inline sixtools/preliminary_processing.py]: Let's see if we can get rid of this slow inner for loop. I think this is equivalent:

```py
for event in raw_data:
    event = np.asarray(event)  # Ensure this is an array (not, say, a list).
    # Select all of 0th axis an a subset of 1st and 2nd axes.
    yield arr[(np.s_[:],) + regions[region]]
```

File: sixtools/preliminary_processing.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# we should change the default for regions in extract_regions to None.
+# Also the manual is really confusing regarding what numbers need to go here,
+# we should discuss this.
+regions = {'dark_low2theta': np.s_[2:1635, 4:1601],
+           'data_low2theta': np.s_[1651:3286, 4:1601],
+           'dark_high2theta': np.s_[2:1635, 4:1601],
+           'data_high2theta': np.s_[1651:3286, 4:1601]}
+
+
+def extract_region(raw_data, region, regions=regions):
```
Comment (tacaswell) [inline sixtools/preliminary_processing.py]: It is very bad practice to have mutable defaults.  They are bound at function definition time and if you mutate them it persist 

```python
def surprise(inp=[0]):
    inp[0] += 1
    print(inp)
```

Comment (tacaswell) [reply sixtools/preliminary_processing.py]: The singular name vs plural behavior is also a bit surprising.

File: sixtools/preliminary_processing.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    --------
+    This is intended to be used in the following, one line, call with XXXX
+    being a scan_id or equivalent (start.rixscam.regions is suppose to point
+    to the metadata containing the regions for this scan):
+    >>> extract_region(db[XXXX].data, region, db[XXXX].start.rixscam.regions)
+    '''
+    # step through each of the 'events' ('yields') in raw_data
+    for event in raw_data:
+        # extract out each array in the list
+        out_list = []
```
Comment (tacaswell) [inline sixtools/preliminary_processing.py]: The region information comes in via a dictionary (which has names) and goes out in a list (which does not).  Either things should go out as a dictionary (to keep the names around) or come in a list (as the names don't matter).

File: sixtools/preliminary_processing.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
+
+# This dictionary should eventually disappear, we should have this region
+# dictionary saved as a configuration attribute of rixscam. At this point
+# we should change the default for regions in extract_regions to None.
+# Also the manual is really confusing regarding what numbers need to go here,
+# we should discuss this.
+regions = {'dark_low2theta': np.s_[2:1635, 4:1601],
```
Comment (tacaswell) [inline sixtools/preliminary_processing.py]: The numbers for high and low are the same, that seems odd...

**Comment by awalter-bnl** [comment]
@mpmdean is this used in your code? If not then I will close the PR, I want to avoid having unused code in the repo.

**Comment by mpmdean** [comment]
@awalter-bnl I copied the [slice, slice] structure, but beyond that, I don't plan to include this, so I think it makes sense to close.

**Comment by awalter-bnl** [comment]
There is no longer a plan to use this so closing the PR

**Pull request closed without merging**